src/main/kotlin/io/emeraldpay/dshackle/upstream/Selector.kt

```diff
@@ -258,10 +301,20 @@ class Selector {
     }
 
     class OrMatcher(val matchers: Collection<LabelSelectorMatcher>) : LabelSelectorMatcher() {
+        private var cause: String = ""
         override fun matches(labels: UpstreamsConfig.Labels): Boolean {
-            return matchers.any { matcher -> matcher.matches(labels) }
+            return matchers.any { matcher ->
+                val matched = matcher.matches(labels)
+                if (!matched) {
+                    cause = matcher.cause()
```

Comment:
I don't think that this version is good. I think you should declare a method like matchesWithCause which always returns success or error, than im abstract class implement matches which internally calls matchesWithCause and strips cause